File: leetcode/0090.go
package leetcode

import "sort"

func subsetsWithDup(nums []int) [][]int {
	ans, tmp, n := make([][]int, 0), make([]int, 0), len(nums)
	sort.Ints(nums)
	var dfs func(int)
	dfs = func(lens int) {
		ans = append(ans, append([]int{}, tmp...))
		if lens == n {
			return
		}
		for j := lens; j < n; j++ {
			if j > lens && nums[j] == nums[j-1] {
				continue
			}
			tmp = append(tmp, nums[j])
			dfs(j + 1)
			tmp = tmp[:len(tmp)-1]
		}
	}
	dfs(0)
	return ans
}

File: leetcode/0090.cpp
class Solution {
public:
    vector<int> t;
    vector<vector<int>> ans;

    void dfs(bool choosePre, int cur, vector<int> &nums) {
        if (cur == nums.size()) {
            ans.push_back(t);
            return;
        }
        dfs(false, cur + 1, nums);
        if (!choosePre && cur > 0 && nums[cur - 1] == nums[cur]) {
            return;
        }
        t.push_back(nums[cur]);
        dfs(true, cur + 1, nums);
        t.pop_back();
    }

    vector<vector<int>> subsetsWithDup(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        dfs(false, 0, nums);
        return ans;
    }
};